openwisp_controller/geo/api/views.py

```diff
@@ -16,11 +21,25 @@
 Device = load_model('config', 'Device')
 Location = load_model('geo', 'Location')
 DeviceLocation = load_model('geo', 'DeviceLocation')
+FloorPlan = load_model('geo', 'FloorPlan')
 
 
 class DevicePermission(BasePermission):
     def has_object_permission(self, request, view, obj):
-        return request.query_params.get('key') == obj.key
+        if request.query_params.get('token'):
+            received_token = request.query_params.get('token')
+            user_token = request.user.auth_token.key
+            if received_token == user_token:
+                return True
+        elif request.query_params.get('key'):
+            received_key = request.query_params.get('key')
+            try:
+                device_key = obj.key
+            except AttributeError:
+                device_key = obj.device.key
+            if received_key == device_key:
+                return True
```

Comment:
There has been a misunderstanding here.

This is not what I meant. I meant the API should work normally as all other endpoints do, using `SessionAuthentication` and `TokenAuthentication` and in addition the key can be used as well to maintain backward compatibility (because this endpoint can be used by devices to update their position).

I think we have two options:

- if doable, we can convert this class into a [DRF authentication class (see the provided link for more info)](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py), but I am not sure we have access to the object in the authentication class, if we do, we can go this route
- alternatively we can add a new authentication class which just checks for the presence of `request.query_params.get('key')`, then in this class, we should change it so that we enforce the key to be correct only if the request comes from an unauthenticated user

However, we also need permissions to be consistent, with the key a request can access the API no matter what, but an authenticated user must have the required django permissions set and the APi must respect multi-tenancy (users should only be able to access items related to their organization), which doesn't seem to be the case here, so with the current implementation any authenticated user would be able to edit any location.

Reply:
@nemesisdesign What we want can be done in this way:- I am changing the permission dynamically based on authentication of a user, have a look
``` python
class DevicePermission(BasePermission):
    def has_object_permission(self, request, view, obj):
        if request.query_params.get('key'):
            received_key = request.query_params.get('key')
            try:
                device_key = obj.key
            except AttributeError:
                device_key = obj.device.key
            return received_key == device_key
        else:
            return False

class DeviceLocationView(
    FilterByOrganizationManaged, generics.RetrieveUpdateDestroyAPIView
):
    serializer_class = DeviceLocationSerializer
    authentication_classes = [
        BearerAuthentication,
        SessionAuthentication,
        TokenAuthentication,
    ]
    permission_classes = [IsAuthenticated, DjangoModelPermissions]
    queryset = Device.objects.select_related(
        'devicelocation', 'devicelocation__location'
    )

    def get_organization_queryset(self, qs):
        # Overriding this method because the class
        # `FilterByOrganizationManaged` tries to
        # filter object for non-authenticated users.
        if self.request.user.is_authenticated:
            return qs.filter(**{self.organization_lookup: getattr(self.request.user, self._user_attr)})
        else:
            return qs

    def get_permissions(self):
        if not self.request.user.is_authenticated:
            if self.request.META.get('QUERY_STRING'):
                self.permission_classes.clear()
                self.permission_classes.append(DevicePermission)
        return [perm() for perm in self.permission_classes]
```

Reply:
> alternatively we can add a new authentication class which just checks for the presence of request.query_params.get('key'), then in this class, we should change it so that we enforce the key to be correct only if the request comes from an unauthenticated user

In DRF First authentication is checked and based on it, the permission classes are checked i.e., if the request get's authenticated from the authentication classes then the permissions are decided. we cannot check the authentication of a request inside a authentication class.

Reply:
I am also trying to implement it by `CustomAuthentication` class